### DISCUSSION TITLE
N-S module calculation error, EXIT CODE: 9

### DISCUSSION THREAD

**Discussion by ScouperLH** [idaholab/moose#19460] [Q&A General] [answered]
Hello everyone,
  Accept my warm greetings.
I used to calculate the flow and heat transfer in an "S"-shaped pipe. 
The error is reported as follows.
![dubug](https://user-images.githubusercontent.com/35528725/143035904-34fe21b4-9469-4227-ad55-50bc11e1ddc4.jpg)

My files as follows:
![image](https://user-images.githubusercontent.com/35528725/143037133-90e43489-0f11-4e93-8e04-c665f19433da.png)

```
mu = 0.1
rho = 1
k = .01
cp = 1
vel = 'velocity'
velocity_interp_method = 'rc'
advected_interp_method = 'average'

[Mesh]
  type = FileMesh #Read in mesh from file
  file = channel.e
[]

[Problem]
  fv_bcs_integrity_check = true
[]

[Variables]
  [u]
    type = INSFVVelocityVariable
    initial_condition = 1
  []
  [v]
    type = INSFVVelocityVariable
    initial_condition = 1
  []
  [w]
    type = INSFVVelocityVariable
    initial_condition = 1
  []
  [T]
    type = INSFVEnergyVariable
  []
  [pressure]
    type = INSFVPressureVariable
  []
  [lambda]
    family = SCALAR
    order = FIRST
  []
[]

[ICs]
  [T]
    type = ConstantIC
    variable = T
    value = 300
  []
[]

[AuxVariables]
  [U]
    order = CONSTANT
    family = MONOMIAL
    fv = true
  []
[]

[AuxKernels]
  [mag]
    type = VectorMagnitudeAux
    variable = U
    x = u
    y = v
    z= w
  []
[]


[FVKernels]
  [mass]
    type = INSFVMassAdvection
    variable = pressure
    advected_interp_method = ${advected_interp_method}
    velocity_interp_method = ${velocity_interp_method}
    vel = 'velocity'
    pressure = pressure
    u = u
    v = v
    w= w
    mu = ${mu}
    rho = ${rho}
  []
  [mean_zero_pressure]
    type = FVScalarLagrangeMultiplier
    variable = pressure
    lambda = lambda
  []

  [u_time]
    type = INSFVMomentumTimeDerivative
    variable = 'u'
    rho = ${rho}
 []
  [u_advection]
    type = INSFVMomentumAdvection
    variable = u
    advected_quantity = 'rhou'
    vel = 'velocity'
    advected_interp_method = ${advected_interp_method}
    velocity_interp_method = ${velocity_interp_method}
    pressure = pressure
    u = u
    v = v
    w=w
    mu = ${mu}
    rho = ${rho}
  []
  [u_viscosity]
    type = FVDiffusion
    variable = u
    coeff = ${mu}
  []
  [u_pressure]
    type = INSFVMomentumPressure
    variable = u
    momentum_component = 'x'
    pressure = pressure
  []

  [v_time]
    type = INSFVMomentumTimeDerivative
    variable = v
    rho = ${rho}
  []
  [v_advection]
    type = INSFVMomentumAdvection
    variable = v
    advected_quantity = 'rhov'
    vel = 'velocity'
    advected_interp_method = ${advected_interp_method}
    velocity_interp_method = ${velocity_interp_method}
    pressure = pressure
    u = u
    v = v
    w=w
    mu = ${mu}
    rho = ${rho}
  []
  [v_viscosity]
    type = FVDiffusion
    variable = v
    coeff = ${mu}
  []
  [v_pressure]
    type = INSFVMomentumPressure
    variable = v
    momentum_component = 'y'
    pressure = pressure
  []

  [w_time]
    type = INSFVMomentumTimeDerivative
    variable = w
    rho = ${rho}
  []
 [w_advection]
    type = INSFVMomentumAdvection
    variable = w
    advected_quantity = 'rhow'
    vel = 'velocity'
    advected_interp_method = ${advected_interp_method}
    velocity_interp_method = ${velocity_interp_method}
    pressure = pressure
    u = u
    v = v
    w=w
    mu = ${mu}
    rho = ${rho}
  []
  [w_viscosity]
    type = FVDiffusion
    variable = w
    coeff = ${mu}
  []
  [w_pressure]
    type = INSFVMomentumPressure
    variable = w
    momentum_component = 'z'
    pressure = pressure
  []

  [temp_time]
    type = INSFVEnergyTimeDerivative
    variable = T
    rho = ${rho}
    cp_name = 'cp'
  []
  [temp_conduction]
    type = FVDiffusion
    coeff = 'k'
    variable = T
  []
  [temp_advection]
    type = INSFVEnergyAdvection
    variable = T
    vel = ${vel}
    velocity_interp_method = ${velocity_interp_method}
    advected_interp_method = ${advected_interp_method}
    pressure = pressure
    u = u
    v = v
    w= w
    mu = ${mu}
    rho = ${rho}
  []
[]

[FVBCs]
  [inlet-u]
    type = INSFVInletVelocityBC
    boundary = 'inlet'
    variable = u
    function = '1'
  []
  [inlet-v]
    type = INSFVInletVelocityBC
    boundary = 'inlet'
    variable = v
    function = '0'
  []
 [inlet-w]
    type = INSFVInletVelocityBC
    boundary = 'inlet'
    variable = w
    function = '0'
  []

  [walls-u]
    type = INSFVNoSlipWallBC
    boundary = 'walltop wallbtm wallin wallout'
    variable = u
    function = 0
  []
  [walls-v]
    type = INSFVNoSlipWallBC
    boundary = 'walltop wallbtm wallin wallout'
    variable = v
    function = 0
  []
  [walls-w]
    type = INSFVNoSlipWallBC
    boundary = 'walltop wallbtm wallin wallout'
    variable = w
    function = 0
  []
  [T_hot]
    type = FVDirichletBC
    variable = T
    boundary = 'walltop'
    value = 310
  []

  [T_cold]
    type = FVDirichletBC
    variable = T
    boundary = 'wallbtm'
    value = 300
  []

  [outlet_p]
    type = INSFVOutletPressureBC
    boundary = 'outlet'
    variable = pressure
    function = '0'
  []
[]

[Materials]
  [functor_constants]
    type = ADGenericConstantFunctorMaterial
    prop_names = 'cp k'
    prop_values = '${cp} ${k}'
  []
  [ins_fv]
    type = INSFVMaterial
    u = 'u'
    v = 'v'
    w = 'w'
    pressure = 'pressure'
    temperature = 'T'
    rho = ${rho}
  []
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'
  num_steps = 20
  dt = .5
  dtmin = .5
  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_pc_type -sub_pc_factor_shift_type'
  petsc_options_value = 'asm      200                lu           NONZERO'
  line_search = 'none'
  nl_abs_tol = 1e-10
  nl_max_its = 50
  l_max_its = 200
[]

[Outputs]
  exodus = true
  csv = true
  perf_graph = true # prints a performance report to the terminal
[]
```

I used the command “ mpiexec -n 4”, but  it's also keep slowly.
How can I debug it?

Thanks for your reading.

**Comment by lindsayad** [comment]
Hi, that error message typically indicates that you are out of memory. What is your MOOSE revision? This should be printed shortly above the top of the text that you pasted in here. Also, how much memory is on your machine?

**Reply by ScouperLH** [reply]
HELLO, Thanks for your answer.
My MOOSE revision is git commit d4b8ea64ef on 2021-11-22
![image](https://user-images.githubusercontent.com/35528725/143162123-2f341771-9a16-4f0f-8237-d42ad6967d0d.png).

I use a VMware to run the test, it has 33GB memory,16CPUs. I have tried again, used 16 CPUs，the same error message appeared. 
![image](https://user-images.githubusercontent.com/35528725/143163272-4f511871-81b9-41ba-b4bc-ca907df005ea.png)

![image](https://user-images.githubusercontent.com/35528725/143163732-7b4f6a26-b3cf-456b-9060-310ae3a8aefd.png)
![image](https://user-images.githubusercontent.com/35528725/143163754-ba588aca-8c17-4de6-aa5f-e21343991a8d.png)
![image](https://user-images.githubusercontent.com/35528725/143163773-678762d7-6fed-4a10-91f9-60a73914c75e.png)

**Reply by GiudGiud** [reply]
I think the memory graph shows you are running out of memory

**Reply by ScouperLH** [reply]
You are right, now I will try the following: 
1. Increase the memory of the virtual machine; 
2. Simplify the pipeline; 
3. Try to use the PJFNK method of calculation.
I will report my results later.

**Reply by GiudGiud** [reply]
These are all good ideas. Another two are to rework the mesh and to turn off cell_gradient_caching for each variable. That will slow down the code a bit but it ll cut the memory cost. 
https://mooseframework.inl.gov/source/variables/MooseVariableFV.html

**Reply by lindsayad** [reply]
You may also try using a distributed mesh. I hate to say it but PJFNK will not save you any memory here. We use AD and fill out all the non-zero coupling entries in the preconditioning matrix.

You may consider [memory profiling](https://mooseframework.inl.gov/application_development/profiling.html) and seeing where the highest memory use is. Turning `cell_gradient_caching = false` in all your variable blocks would be a big memory save as @GiudGiud suggested. I'd definitely start there and with a distributed mesh (after increasing your virtual machine's memory)

**Reply by lindsayad** [reply]
Oh I forgot to ask ... what is your `-pc_type`? That will play a significant role in memory usage

**Reply by ScouperLH** [reply]
> Oh I forgot to ask ... what is your `-pc_type`? That will play a significant role in memory usage

It's '-pc_type -ksp_gmres_restart -sub_pc_type -sub_pc_factor_shift_type'.

**Reply by GiudGiud** [reply]
also I would remove this
```
[AuxVariables]
  [U]
    order = CONSTANT
    family = MONOMIAL
    fv = true
  []
[]

[AuxKernels]
  [mag]
    type = VectorMagnitudeAux
    variable = U
    x = u
    y = v
    z= w
  []
[]
```
this is for postprocessing and you can do that operation in paraview easily

**Reply by GiudGiud** [reply]
~~@ScouperLH we need the petsc_values as well not just the names~~
nevermind it s in the input file
  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_pc_type -sub_pc_factor_shift_type'
  petsc_options_value = 'asm      200                lu           NONZERO'

building that Krylov base is for sure expensive. @lindsayad what do you recommend?

**Reply by lindsayad** [reply]
IIRC PETSc does not allocate memory for all the GMRES vectors up front. I believe it allocates every 30 iterations or so as needed. @fdkong could comment on this. I think at this point we are kind of guessing what the big memory contributors are. A heap profile would settle this right away. It looks like these are all MOOSE objects. @ScouperLH how large is your mesh? If it's not too egregiously large, you can send it to alexlindsay239@gmail.com (or I suppose send me a google drive link), and I could profile this problem for you.

**Reply by friedmud** [reply]
> IIRC PETSc does not allocate memory for all the GMRES vectors up front. I believe it allocates every 30 iterations or so as needed. @fdkong could comment on this. I think at this point we are kind of guessing what the big memory contributors are. A heap profile would settle this right away. It looks like these are all MOOSE objects. @ScouperLH how large is your mesh? If it's not too egregiously large, you can send it to alexlindsay239@gmail.com (or I suppose send me a google drive link), and I could profile this problem for you.

It allocates all vectors up front... so with these options it will allocate 200 solution length vectors immediately.

By default the number is 30 - and all 30 are allocated immediately.

**Reply by lindsayad** [reply]
I don't think that's true. You can see the `-ksp_gmres_preallocate` option at the [KSPGMRES](https://petsc.org/release/docs/manualpages/KSP/KSPGMRES.html) manual page. There is this code in `gmres.c`:
```c
  if (gmres->q_preallocate) {
    gmres->vv_allocated = VEC_OFFSET + 2 + max_k;

    ierr = KSPCreateVecs(ksp,gmres->vv_allocated,&gmres->user_work[0],0,NULL);CHKERRQ(ierr);
    ierr = PetscLogObjectParents(ksp,gmres->vv_allocated,gmres->user_work[0]);CHKERRQ(ierr);

    gmres->mwork_alloc[0] = gmres->vv_allocated;
    gmres->nwork_alloc    = 1;
    for (k=0; k<gmres->vv_allocated; k++) {
      gmres->vecs[k] = gmres->user_work[0][k];
    }
  } else {
    gmres->vv_allocated = 5;

    ierr = KSPCreateVecs(ksp,5,&gmres->user_work[0],0,NULL);CHKERRQ(ierr);
    ierr = PetscLogObjectParents(ksp,5,gmres->user_work[0]);CHKERRQ(ierr);

    gmres->mwork_alloc[0] = 5;
    gmres->nwork_alloc    = 1;
    for (k=0; k<gmres->vv_allocated; k++) {
      gmres->vecs[k] = gmres->user_work[0][k];
    }
  }
```
I think unless you specify the `-ksp_gmres_preallocate` option, you're going to get 5 vectors pre-allocated, and then more will be created as needed. The `gmresimpl.h` implies the default preallocation option is 0
```c
#define KSPGMRESHEADER                                                  \
  /* Hessenberg matrix and orthogonalization information. */            \
  PetscScalar *hh_origin;   /* holds hessenburg matrix that has been multiplied by plane rotations (upper tri) */ \
  PetscScalar *hes_origin;  /* holds the original (unmodified) hessenberg matrix which may be used to estimate the Singular Values of the matrix */ \
  PetscScalar *hes_ritz;    /* holds the last full Hessenberg matrix to compute (harmonic) Ritz pairs */ \
  PetscScalar *cc_origin;   /* holds cosines for rotation matrices */   \
  PetscScalar *ss_origin;   /* holds sines for rotation matrices */     \
  PetscScalar *rs_origin;   /* holds the right-hand-side of the Hessenberg system */ \
                                                                        \
  PetscScalar *orthogwork; /* holds dot products computed in orthogonalization */ \
                                                                        \
  /* Work space for computing eigenvalues/singular values */            \
  PetscReal   *Dsvd;                                                    \
  PetscScalar *Rsvd;                                                    \
                                                                        \
                                                                        \
  PetscReal haptol;                                       /* tolerance for happy ending */ \
  PetscInt  max_k;                                        /* number of vectors in Krylov space, restart size */ \
  PetscInt  nextra_vecs;                                  /* number of extra vecs needed, e.g. for a pipeline */ \
                                                                        \
  PetscErrorCode (*orthog)(KSP,PetscInt);                    \
  KSPGMRESCGSRefinementType cgstype;                                    \
                                                                        \
  Vec      *vecs;                                        /* the work vectors */ \
  Vec      *vecb;                                        /* holds the last full basis vectors of the Krylov subspace to compute (harmonic) Ritz pairs */ \
  PetscInt q_preallocate;    /* 0=don't preallocate space for work vectors */ \
  PetscInt delta_allocate;    /* number of vectors to preallocaate in each block if not preallocated */ \
  PetscInt vv_allocated;      /* number of allocated gmres direction vectors */ \
  PetscInt vecs_allocated;                              /*   total number of vecs available */ \
  /* Since we may call the user "obtain_work_vectors" several times, we have to keep track of the pointers that it has returned */ \
  Vec      **user_work;                                              \
  PetscInt *mwork_alloc;       /* Number of work vectors allocated as part of  a work-vector chunck */ \
  PetscInt nwork_alloc;        /* Number of work vector chunks allocated */ \
                                                                        \
  /* Information for building solution */                               \
  PetscInt    it;              /* Current iteration: inside restart */  \
  PetscInt    fullcycle;       /* Current number of complete cycle */ \
  PetscScalar *nrs;            /* temp that holds the coefficients of the Krylov vectors that form the minimum residual solution */ \
  Vec         sol_temp;        /* used to hold temporary solution */ \
  PetscReal   rnorm0;          /* residual norm at beginning of the GMRESCycle */ \
  PetscReal   breakdowntol;    /* A relative tolerance is used for breakdown check in GMRESCycle */

typedef struct {
  KSPGMRESHEADER
} KSP_GMRES;
```

**Reply by lindsayad** [reply]
Yea this is in `KSPSetFromOptions_GMRES`
```c
  flg  = PETSC_FALSE;
  ierr = PetscOptionsBool("-ksp_gmres_preallocate","Preallocate Krylov vectors","KSPGMRESSetPreAllocateVectors",flg,&flg,NULL);CHKERRQ(ierr);
  if (flg) {ierr = KSPGMRESSetPreAllocateVectors(ksp);CHKERRQ(ierr);}
```

**Reply by fdkong** [reply]
We do not really allocate all vectors up front.

1. We allocate about 5 vectors during the setup phase. 2 of them are used as the user work spaces, and 3 of them are used to store  Krylov subspace vectors. 

2. When we have more than three iterations, a delta number (10 by default) of vectors will be allocated. This step will repeat until we reach the restart number, e.g. it is 30 by default but it is 200 in the mentioned case (-ksp_gmres_restart 200)

3. The allocated vectors will be kept into the second/ third nonlinear iterations and so on. We won't release  memory from one nonlinear to another nonlinear iteration

**Comment by ScouperLH** [comment]
> These are all good ideas. Another two are to rework the mesh and to turn off cell_gradient_caching for each variable. That will slow down the code a bit but it ll cut the memory cost. https://mooseframework.inl.gov/source/variables/MooseVariableFV.html

Thank you, I will try your suggestion.

And as @lindsayad says, I use PJFNK to calculate, but it not save the memory. Also thanks for your answer!@lindsayad

**Reply by GiudGiud** [reply]
We've used AD in the finite volume for Navier Stokes. So we have a perfect numerical Jacobian and you might as well use it with Newton instead of PJFNK. Only reason to use PJFNK would be for memory imo, or if some other parts of the model didnt use AD

**Reply by lindsayad** [reply]
@GiudGiud the issue is that with AD we fill the whole matrix (we don't respect users request for block diagonal or other custom sparsity pattern), so PJFNK and NEWTON will have the same memory requirements. So given that, indeed you should always use NEWTON when you are using AD as it will consume the same amount of memory and likely be way faster

**Comment by friedmud** [comment, marked as answer]
You are using LU - sure... it's only on the parallel partitions with ASM... but LU will still take quite a lot of memory.  Try switching to something else like ILU with a small amount of fill to save even more RAM.

**Reply by fdkong** [reply]
This is right. If we do not make any mistake, LU memory will dominate the whole simulation. It will be helpful if you could tell us, DoFs and Number of elements/nodes

**Reply by ScouperLH** [reply]
Ok,it's the DoF, nodes and elements.
![image](https://user-images.githubusercontent.com/35528725/143358682-77a6ea1e-153d-44d5-a025-df2986926c28.png)

**Reply by fdkong** [reply]
Thanks for information.  You have about half million DoFs, and you could run out of memory with ASM (LU) if the machine does not have enough memory.

How much memory you have on machine you were running? How much memory per core?

**Reply by ScouperLH** [reply]
When I first calculated, my VMware is 33GB, 16cores, and per core has 2GB. Then I changed the memory to 50GB, but the memory is still not enough, which is unreasonable for a geometric model of this scale. I imported the grid into Fluent and set up a simple laminar flow model calculation to confirm that there is no problem with the grid. Finally, I simplified the geometric model, and I will report to you after the post-processing confirms that the result is correct.

**Reply by fdkong** [reply]
Just for debugging. Could you change your input file to take this? As Derek mentioned earlier, to replace LU with ILU to save memory.


```
  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_pc_type -sub_pc_factor_shift_type'
  petsc_options_value = 'asm      200                ilu           NONZERO' 
```

If this won't work, we might actually have some memory issue somewhere. In that case, it would be really helpful if you would share an example with us,  and we will try to nail down the issue

**Reply by ScouperLH** [reply]
After modificated ILU it worked, I'm uploading my files to share you later.

**Reply by fdkong** [reply]
That meant you should select a memory-efficient solver. If ILU works for you, you could keep using that. Linear solver/preconditioner and nonlinear solver are problem-dependent and sometimes tricky. This might improve in the future as the team is working on a dedicated CFD solver strategy.

**Comment by ScouperLH** [comment]
Hello, moose experts, I have made the following modifications and tests:
1. Modify the variable cache and delete auxiliary variables.@GiudGiud  Run again and the same error message appears.
At the same time, the PJFNK method does not reduce memory consumption, which has been described in the previous comments.@lindsayad


2. Increase the memory to 64GB and redraw the geometric model and grid blend&channel/blend.e (in fact, new elems&nodes did not decrease, they increased, because I mistook the  grid number of the first model. The skewness of the new grid is better than that of the old grid.  They are in https://github.com/ScouperLH/bflow2.git )

Results: the new grid run in the same inputfile (channel-3d2. i). It should be emphasized that the Lu command used by the solver uses 54gb of memory. Finally, the result is calculated. At present, I have no post-processing results, because I am still learning paraview, and may try to use Tecplot. 

3. Following suggestion of @friedmud @fdkong , I modified the solver command to ILU, which can successfully calculate the input files of blend model and channel model, and the memory usage is about 20GB.

4. BUT, when I used ILU calculate channel-3d1st2.i  (with channel.e)，new error appears:

![image](https://user-images.githubusercontent.com/35528725/143412327-abb9cc0d-c56d-4884-a57f-a34dd87ae8ee.png)

**Reply by ScouperLH** [reply]
I just recalculated this file, modified the parameters according to [gambka 's comment](https://github.com/idaholab/moose/issues/18819)， the error disappeared, and got the result.

**Reply by GiudGiud** [reply]
how big was the latest time step? I m going to fix that

**Reply by GiudGiud** [reply]
the one that failed that is

**Reply by ScouperLH** [reply]
> how big was the latest time step? I m going to fix that

Time Step 20, time = 10, dt = 0.5
![20211125204835](https://user-images.githubusercontent.com/35528725/143444805-b1dd38db-1b16-4b27-b49c-e2f5da5c0ce8.png)


When it calculate in Step 19, error occur. It was very close to the solution.

**Reply by GiudGiud** [reply]
Umm ok so what did you change?

The comment you linked is related to the time stepper just missing it s end time by a tiny fraction, and the solver dying when trying to solve for 1e-14 s .  I m looking for the magnitude of the tiny step it took in your case, if that was the problem

**Reply by ScouperLH** [reply]
I change the nl_abs_tol=1.1-10

**Reply by GiudGiud** [reply]
oh ok that's not the same thing then. 

btw it usually helps to scale your non-linear variables to reduce the residual / help convergence. You can do that manually for each variable or use automatic scaling. some docs:
auto scaling: https://mooseframework.inl.gov/newsletter/2019_07.html
manual scaling: https://mooseframework.inl.gov/modules/porous_flow/convergence.html

**Reply by ScouperLH** [reply]
Ok, I will read the files
